### ISSUE TITLE
Pedersen hash windowing performance

### ISSUE THREAD

**Issue by imeckler** [zcash/zcash#3899]
Is there an explanation somewhere of the reasoning behind the `g, 16 g, 16^2 g, …` version of the Pedersen hash function? (The alternative being just to use independent generators.) 

From my calculations, window-tables for the scalar-packing version seems like they consume more memory (for the same performance improvement) than window-tables for the independent generator version.

**Comment by daira** [comment]
[All theorem/section references in this comment are from the [Sapling protocol spec](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf).]

This is to ensure that the premises of Theorems A.3.4, A.3.5 and 5.4.1 are satisfied. The reasoning goes as follows:

It's faster to add points on the Montgomery curve than on the Edwards curve (just because the equations are simpler). But the downside is that we must prove that the input points of each addition have distinct x-coordinates.

How do we do that? Well, informally, Theorem A.3.4 allows us to go from distinctness (disregarding sign) of the scalars k by which we multiply any given generator G, to distinctness of the x-coordinates of the corresponding points [k] G. (Even more informally: this works because half of the scalars multiply to half of the points, and then the other half is obtained by negating each scalar and point, where negation affects only the y-coordinate.)

Thus, controlling the relation between scalars applied to the same generator is essentially the only tool we have to ensure that the x-coordinates are distinct. If we were to use independent generators for each window, then when adding the resulting points we would have to use Edwards addition, since we have no way to prove correctness of Montgomery addition in that case.

Let's see how much difference this makes in practice. Assume that the input is already boolean-constrained.

Option A: use Edwards addition exclusively, looking up Edwards points in the window tables. We still use the negation trick to obtain the same 3-bit effective window size with a 1-bit smaller lookup and a conditional negation. The cost per 3-bit window is 1C to compute s<sub>&</sub> for the lookup, 1C for the conditional negation of the y-coordinate, and 6C for the Edwards addition. The lookup itself is linear and so free. So the cost per bit is 8/3 C.

Option B (as in Sapling): use Montgomery addition as much as possible, only using Edwards when we overflow the number of bits available in a scalar. There are few Edwards additions and Montgomery→Edwards conversions compared to the number of bits, so to see why this option is cheaper, we can neglect those (section A.3.3.9 has the full cost analysis). The cost per 3-bit window is 1C to compute s<sub>&</sub> for the lookup, 1C for the conditional negation of the y-coordinate, and 3C for the Montgomery addition. Again, the lookup is linear and so free. So the cost per bit is roughly 5/3 C.

You can consider other window sizes for each option, but they don't give any improvement as far as I know.

Note that I don't make any claim of optimality for this approach. It's entirely possible that other improvements are possible. We can brainstorm them on this ticket if you would like.

**Comment by daira** [comment]
Finally, the reason why we use base 16 rather than base 8 for the scalar representation is that base 8 could cause collisions when using signed digits (e.g. 1 \* 8 + (-4) = 4, so (1, -4) would collide with (0, 4)). Using unsigned digits would require either a more expensive lookup or smaller windows. See Theorem A.3.5 for why that works, but it's rather ugly and there might be other ways of doing it.

**Comment by daira** [comment]
> From my calculations, window-tables for the scalar-packing version seems like they consume more memory (for the same performance improvement) than window-tables for the independent generator version.

I'm a bit confused by this. I assume you mean window tables outside the circuit (since the ones in the circuit are "wired in"). There are at least four possible implementations with different time/memory tradeoffs: for each generator G,

i) precompute [k \* 16<sup>j</sup>] G for k ∈ {1..4} and j ∈ {1..63}
ii) precompute [k \* 16<sup>j</sup>] G for k ∈ {1, 3} and j ∈ {1..63};
iii) precompute only [16<sup>j</sup>] G for j ∈ {1..63};
iv) just store G.

(There's no advantage in storing points for negative k, because negation is so cheap.)

Assuming table entries in affine representation (64 bytes each), the storage cost *per generator* of each implementation for Option B is:

i) 4\*63\*64 bytes (15.75 KiB);
ii) 2\*63\*64 bytes (~7.88 KiB);
iii) 63\*64 bytes (~3.94 KiB);
iv) 64 bytes.

In Option B, each 189 bits of the maximum input size requires another generator. In Option A, implementations i) to iii) would each have the same storage requirement for the same maximum input size, but implementation iv) would be ruled out. (For low-memory implementations you could store just the index i for FindGroupHash and compute GroupHash on the fly, I suppose.)

**Comment by ebfull** [comment]
There is perhaps another (accidental) benefit, which is that because computing the exponents is cheap and there are few group elements, the Pedersen hash can be computed outside of the circuit using fast multi-exponentiation algorithms. If every lookup was a distinct base this would not be as efficient I think. (Could be wrong.)

But yes, the primary motivation was avoiding a difficult security argument/proof. What do you save anyway, a couple dozen constraints or something?

**Comment by daira** [comment]
If you depend on x-coordinate collisions not occurring, then you end up with 5·c - 1 constraints instead of 5·c + 5·n - 6 constraints. This saves 5·(n+1) constraints where n = ceiling(ℓ/189), for input size ℓ bits. For example, with ℓ = 516 bits you save 20 constraints out of 869. It's not worth it.

**Comment by imeckler** [comment]
Sorry for not explaining properly. To be clear I was only talking about outside the SNARK. Also I had not considered the possibility of using a multi-exp algorithm. Let me clarify what I meant as follows:

Let's say we want to hash n triples and our scalar field size is r bits. There are two approaches I was considering.

1. Do the scalar-packing thing with one generator every roughly r / 4 triples. The hash algorithm is pack triples into the scalar until we hit r / 4 triples, then do an exponentiation with a window table and accumulate the result into a group element. 

The memory usage for a window of size k is (n / (r / 4)) * (r / k) * 2^k = (4n / k) * 2^k. The number of group operations is lower bounded by (n / (r / 4)) exponentiations, each of which does roughly (r / k) adds, so (4n / k) adds at least in total.

2. Just accumulate into a group element, but for every m bits, have a window table indexed by sequences of triples of length m. Do a lookup in each window, then add all the results together.

Basically the map (2^3)^n -> ((2^3)^m)^(n / m) -> G^(n / m) -> G.

The memory usage here is 2^(3m) * (n / m) and the number of group operations is basically (n / m).

---

Let's compare these two approaches. Let's say we want to make them have the same speed, so we'll have

4 n / k = n / m

which implies k = 4 m. Now we take a look at the memory usage. Approach 1 requires

(4 n / k) * 2^k = (4 n / 4 m) * 2^(4 m) = (n / m) * 2^(4 m)

which is worse than the (n / m) * 2^(3m) required by approach 2. In other words, if we want approach 1 to be as fast, then we need to use more memory than in approach 2.

**Comment by daira** [comment]
@imeckler wrote:
> Sorry for not explaining properly.

No problem.

> To be clear I was only talking about outside the SNARK. Also I had not considered the possibility of using a multi-exp algorithm. Let me clarify what I meant as follows:
> 
> Let's say we want to hash n triples and our scalar field size is r bits. There are two approaches I was considering.
> 
> 1. Do the scalar-packing thing with one generator every roughly r / 4 triples. The hash algorithm is pack triples into the scalar until we hit r / 4 triples, then do an exponentiation with a window table and accumulate the result into a group element. This is I believe what is used in zcash.

Yes.

> The memory usage for a window of size k is (n / (r / 4)) * (r / k) * 2^k = (4n / k) * 2^k.

Remember that they're signed digits so you only need to store 2<sup>k-1</sup> entries per window. Also, n is the number of triples, so it's n \* 2<sup>k-1</sup> entries. I'm not sure where you get the 4/k factor from.

(This is option i) in my comment above. As I said, there are several lower-memory options for the same hash, and there are also higher-memory approaches as you point out.)

> The number of group operations is lower bounded by (n / (r / 4)) exponentiations, each of which does roughly (r / k) adds, so (4n / k) adds at least in total.

It's only n-1 additions to compute a hash.

> 1. Just accumulate into a group element, but for every m triples, have a window table indexed by sequences of triples of length m.

Again you can halve the size of this table by taking advantage of negation.

> Do a lookup in each window, then add all the results together.
> 
> Basically the map (2^3)^n -> ((2^3)^m)^(n / m) -> G^(n / m) -> G.

You can do exactly the same thing to combine window tables [k \* 16<sup>j</sup>] G for adjacent j, with the same memory usage and speed.

**Comment by imeckler** [comment]
Thank you for the reply, it's a good point that any optimization you can do for the independent generators hash you can do for the 'scalar packing' hash.

> It's only n-1 additions to compute a hash.

I was just referring to (I believe from my reading the code a few weeks ago) the algorithm implemented in zcash, which I think works as described. This algorithm does (n / (r/4)) exponentiations, each of which does roughly r / k adds, so 4n / k adds in total.

Great point about halving the table size.

**Comment by daira** [comment]
@imeckler wrote:
> I was just referring to (I believe from my reading the code a few weeks ago) the algorithm implemented in zcash, which I think works as described. This algorithm does (n / (r/4)) exponentiations, each of which does roughly r / k adds, so 4n / k adds in total.

I'm not quite sure why we use that algorithm, but in any case, computing the hash outside the circuit is not a performance bottleneck AFAIK.

**Comment by imeckler** [comment]
Got it, thank you for the patient and in-depth responses.

**Comment by ioptio** [comment]
This looks resolved! Feel free to reopen if you have related questions still.
